lib/ruby_lsp/requests/document_highlight.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module RubyLsp
+  module Requests
+    # The [document highlight](https://microsoft.github.io/language-server-protocol/specification#textDocument_documentHighlight)
+    # informs the editor all relevant elements of the currently pointed item for highlighting. For example, when
+    # the cursor is on the `F` of the constant `FOO`, the editor should identify other occurences of `FOO`
+    # and highlight them.
+    #
+    # For writable elements like constants or variables, their read/write occurrences should be highlighted differently.
+    # This is achieved by sending different "kind" attributes to the editor (2 for read and 3 for write).
+    #
+    # # Example
+    #
+    # ```ruby
+    # FOO = 1 # should be highlighted as "write"
+    #
+    # def foo
+    #   FOO # should be highlighted as "read"
+    # end
+    # ````
+    class DocumentHighlight < BaseRequest
+      def self.run(document, position)
+        new(document, position).run
+      end
+
+      def initialize(document, position)
+        @highlights = []
+        position = Document::Scanner.new(document.source).find_position(position)
+        @target = find(document.tree, position)
+
+        super(document)
+      end
+
+      def run
+        # no @target means the target is not highlightable
+        return [] unless @target
+
+        visit(@document.tree)
+        @highlights
+      end
+
+      def visit_var_field(node)
```

Comment:
Do we need to call `super` in these visitors so that their children can be visited as well? I would assume we could have nested `VarField` and `VarRef`s within each other, and if we don't call super, then we would not be able to match the nested ones.